src/main/java/com/google/devtools/build/lib/remote/common/RemoteActionExecutionContext.java

```diff
@@ -13,37 +13,71 @@
 // limitations under the License.
 package com.google.devtools.build.lib.remote.common;
 
+import build.bazel.remote.execution.v2.ExecuteResponse;
 import build.bazel.remote.execution.v2.RequestMetadata;
 import com.google.devtools.build.lib.actions.ActionExecutionMetadata;
 import com.google.devtools.build.lib.actions.Spawn;
 import javax.annotation.Nullable;
 
 /** A context that provide remote execution related information for executing an action remotely. */
-public interface RemoteActionExecutionContext {
-  /** The type of the context. */
-  enum Type {
-    REMOTE_EXECUTION,
-    BUILD_EVENT_SERVICE,
+public class RemoteActionExecutionContext {
+  /** The current step of the context. */
+  public enum Step {
+    INIT,
+    CHECK_ACTION_CACHE,
+    UPLOAD_INPUTS,
+    EXECUTE_REMOTELY,
+    UPLOAD_OUTPUTS,
+    DOWNLOAD_OUTPUTS,
+    UPLOAD_BES_FILES,
   }
 
-  /** Returns the {@link Type} of the context. */
-  Type getType();
+  private final Spawn spawn;
+  private final RequestMetadata requestMetadata;
+  private final NetworkTime networkTime;
+
+  @Nullable private ExecuteResponse executeResponse;
+  private Step step;
+
+  public RemoteActionExecutionContext(
+      Spawn spawn, RequestMetadata requestMetadata, NetworkTime networkTime) {
```

Comment:
Same as above.

Reply:
Also - should we make the ctor private to enforce construction through one of the `create()` methods?

Reply:
Done.